tools/c7n_azure/c7n_azure/resources/network_security_group.py

```diff
@@ -52,223 +53,204 @@ class SecurityRuleFilter(Filter):
 
-class RulesAction(BaseAction):
-    """
-    Action to perform on SecurityRules within a Network Security Group
-    """
+        if EXCEPT_PORTS in self.data:
+            if pattern.match(self.data[EXCEPT_PORTS]) is None:
+                raise PolicyValidationError("exceptPorts string has wrong format.")
+        return True
+
+    def build_ports_string(self, nsg, direction_key, ip_protocol):
+        # Build list of ports for a given nsg, True if allow, False if Deny
+        nsg_ports = PortsRangeHelper.build_ports_array(nsg, direction_key, ip_protocol)
+        nsg_ports = [False if x is None else x for x in nsg_ports]
+
+        access = StringUtils.equal(self.access_action, "allow")
+
+        # Find ports with different access level from NSG and this action
+        diff_ports = sorted([p for p in self.action_ports if nsg_ports[p] != access])
+
+        # diff_ports empty means this NSG already satisfies action conditions
+        if len(diff_ports) == 0:
+            return ""
+
+        # Transform diff_ports list to the ranges list
+        first = 0
+        result = []
+        for it in range(1, len(diff_ports)):
+            if diff_ports[first] == diff_ports[it] - (it - first):
+                continue
+            result.append((diff_ports[first], diff_ports[it - 1]))
+            first = it
+
+        # Update tuples with strings, representing ranges
+        result.append((diff_ports[first], diff_ports[-1]))
+        result = [str(x[0]) if x[0] == x[1] else "%i-%i" % (x[0], x[1]) for x in result]
+
+        return result
 
     def process(self, network_security_groups):
 
+        ip_protocol = self.data.get(IP_PROTOCOL, '*')
+        direction = self.data[DIRECTION]
+        # Build a list of ports described in the action.
+        ports = PortsRangeHelper.get_ports_set_from_string(self.data.get(PORTS, '0-65535'))
+        except_ports = PortsRangeHelper.get_ports_set_from_string(self.data.get(EXCEPT_PORTS, ''))
+        self.action_ports = ports.difference(except_ports)
+
         for nsg in network_security_groups:
             nsg_name = nsg['name']
             resource_group = nsg['resourceGroup']
-            for rule in nsg['properties']['securityRules']:
-                self.manager.log.info("Updating access to '%s' for security rule "
-                                      "'%s' in resource group '%s'",
-                                      self.access_action, rule['name'], resource_group)
-                rule['properties']['access'] = self.access_action
-                self.manager.get_client().security_rules.create_or_update(
-                    resource_group,
-                    nsg_name,
-                    rule['name'],
-                    rule
-                )
+
+            # Get list of ports to Deny or Allow access to.
+            ports = self.build_ports_string(nsg, direction, ip_protocol)
+            if ports == '':
```

Comment:
Recommendation - empty string are falsey in python so you can do `if not ports` here. It's more readable, but this works too :)